Hardware tutorial (#330)
XanaduAI/strawberryfields: doc/introduction/remote.rst

.. _starship:

Executing programs on photonic hardware
=======================================

In this tutorial, we demonstrate how Strawberry Fields can execute
quantum programs on remote photonic hardware, via the Xanadu cloud platform. Topics
covered include:

* Configuring Strawberry Fields to access the cloud platform,
* Using the :class:`~.RemoteEngine` to execute jobs remotely (e.g., on a quantum chip),
* Managing submitted jobs,
* Compiling quantum programs for specific quantum chip architectures,
* Saving and running Blackbird scripts, an assembly language for quantum photonic hardware, and
* Embedding and sampling from graphs on photonic chips.

.. warning::

    An authentication token is required to access the Xanadu cloud platform. In the

>     following, replace ``AUTHENTICATION_TOKEN`` with your personal API token.

Author: ```suggestion
    An authentication token is required to access the Xanadu cloud platform. In the
    following, replace ``AUTHENTICATION_TOKEN`` with your personal API token.
    
    If you do not have an authentication token, you can request hardware access via email
    at sf@xanadu.ai.
```

        ops.MeasureFock() | q

Finally, we create the engine. Similarly to the :class:`~.LocalEngine`, the :class:`~.RemoteEngine`

> is in charge of compiling and executing programs, however it differs in that the program will be

Reviewer: ```suggestion
is in charge of compiling and executing programs. However, it differs in that the program will be
```

executed on *remote* devices, rather than on local simulators.

>>> eng = RemoteEngine("X8_01")

> We run the engine by calling ``eng.run``, and pass it the program we

Reviewer: Better to phrase as running the program rather than running the engine

Author: Rephrase this to:
> We can now run the program by calling ``eng.run``, and passing the program to be executed
as well as additional runtime options.

       [0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 1, 1, 0, 2, 1, 2],
       [2, 0, 1, 0, 1, 0, 0, 0]])

The samples returned correspond to 20 measurements (or shots) of the 8 mode quantum program
above. While the majority of modes have measured zero photons, we can see that a couple have
detected single photons, with a few even detecting 2 or 3.

By taking the average of the returned array along the shots axis, we can estimate the
mean photon number of each mode:

>>> np.mean(results.samples, axis=0)
array([0.4 , 0.1 , 0.15, 0.05, 0.3 , 0.3 , 0.45, 0.35])

We can also use the Python collections module to convert the samples into
counts:

>>> from collections import Counter
>>> bitstrings = [tuple(i) for i in results.samples]
>>> counts = {k:v for k, v in Counter(bitstrings).items()}
>>> counts[(0, 0, 0, 0, 0, 0, 0)]

> 2

Author: Yep, people were asking how to do it.

> 2

Reviewer 2: Ok, if it helps people :smile:

.. note::

    The :func:`~.operation` decorator allows you to create your own Strawberry Fields
    operation. This can make it easier to ensure that the same unitary is always
    applied to the signal and idler modes.

    .. code-block:: python3

        from strawberryfields.utils import operation

        @operation(4)
        def unitary(q):
            ops.Interferometer(U) | q
            ops.BSgate(0.543, 0.123) | (q[2], q[0])

        prog = sf.Program(8)

        with prog.context as q:
            ops.S2gate(1.0) | (q[0], q[4])
            ops.S2gate(1.0) | (q[1], q[5])
            ops.S2gate(1.0) | (q[2], q[6])
            ops.S2gate(1.0) | (q[3], q[7])

            unitary() | q[:4]

>             unitary() | q[4:]

Reviewer 2: ```suggestion
            unitary() | q[4:]
 
            ops.MeasureFock() | q
```

    Refer to the :func:`~.operation` documentation for more details.

Managing submitted jobs
-----------------------

In the above example, ``eng.run()`` is a **blocking** method; Python will wait until
the job result has been returned from the cloud before further lines
will execute.

> **Non-blocking jobs** can also be submitted using the :meth:`eng.run_async() <~.RemoteEngine.run_async>` method. Unlike ``eng.run()``, this method returns a :class:`~.Job`

Author: Ah, thanks for catching this Tom! I had a stray `~`, removing this makes it work.

> object that can be queried to get the jobs status.

Reviewer 2: ```suggestion
object that can be queried to get the job's status.
```

>>> job = engine.run_async(program, shots=1)
>>> job.id
"e6ead866-04c9-4d48-ba28-680e8639fc41"
>>> job.status
"queued"

If the job result is not yet available, an ``InvalidJobOperationError`` will be raised:

>>> job.result
InvalidJobOperationError

To check when the results are ready, the job can be refreshed, and the status
checked:

>>> job.refresh()
>>> job.status
"complete"
>>> job.result
[[0 1 0 2 1 0 0 0]]

> Finally, an incomplete job can be *cancelled* by calling :meth:`job.cancel() <~.Job.cancel>`.

Reviewer 2: Although, the method link didn't work for me again.

> Finally, an incomplete job can be *cancelled* by calling :meth:`job.cancel() <~.Job.cancel>`.

Author: Fixed!

Hardware compilation
--------------------

When creating a quantum program to run on hardware, Strawberry Fields allows any

> combination of the following gates when constructing the unitary to be applied:

Reviewer 2: ```suggestion
When creating a quantum program to run on hardware, Strawberry Fields can compile
any collection of the following gates into a multi-mode unitary:
```
Just felt this sentence read a bit strangely.

* `General beamsplitters <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.BSgate.html>`_ (:class:`~.ops.BSgate`),

* `Mach-Zehnder interferometers <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.MZgate.html>`_ (:class:`~.ops.MZgate`), or

* `rotations/phase shifts <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.Rgate.html>`_ (:class:`~.ops.Rgate`).

Furthermore, several automatic decompositions are supported:

* You can use the :class:`~.ops.Interferometer` command to directly pass a
  unitary matrix to be decomposed and compiled to match the device architecture.
  This performs a rectangular decomposition using Mach-Zehnder interferometers.

* You can use :class:`~.ops.BipartiteGraphEmbed` to embed a bipartite graph on
  the GBS chip. Note, however, that the decomposed squeezing values depends on the graph
  structure, so only bipartite graphs that result in equal squeezing on all

>   modes can currently be executed on currently available chips.

Reviewer 2: ```suggestion
  modes can be executed on currently available chips.
```

       [0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 1, 1, 0, 2, 1, 2],
       [2, 0, 1, 0, 1, 0, 0, 0]])

The samples returned correspond to 20 measurements (or shots) of the 8 mode quantum program

> above. While the majority of modes have measured zero photons, we can see that a couple have

Reviewer: ```suggestion
above. Some modes have measured zero photons, and others have
```

detected single photons, with a few even detecting 2 or 3.

By taking the average of the returned array along the shots axis, we can estimate the
mean photon number of each mode:

>>> np.mean(results.samples, axis=0)

> array([0.4 , 0.1 , 0.15, 0.05, 0.3 , 0.3 , 0.45, 0.35])

Reviewer: ```suggestion
array([0.4, 0.1, 0.15, 0.05, 0.3, 0.3, 0.45, 0.35])
```

We can also use the Python collections module to convert the samples into
counts:

>>> from collections import Counter
>>> bitstrings = [tuple(i) for i in results.samples]
>>> counts = {k:v for k, v in Counter(bitstrings).items()}
>>> counts[(0, 0, 0, 0, 0, 0, 0)]
2

.. note::

    The :class:`~.operation` decorator allows you to create your own Strawberry Fields
    operation. This can make it easier to ensure that the same unitary is always
    applied to the signal and idler modes.

    .. code-block:: python3

        from strawberryfields.utils import operation

        @operation(4)
        def unitary(q):
            ops.Interferometer(U) | q
            ops.BSgate(0.543, 0.123) | (q[2], q[0])
            ops.Rgate(0.453) | q[1]
            ops.MZgate(0.65, -0.54) | (q[2], q[3])

        prog = sf.Program(8)

        with prog.context as q:
            ops.S2gate(1.0) | (q[0], q[4])
            ops.S2gate(1.0) | (q[1], q[5])
            ops.S2gate(1.0) | (q[3], q[7])

>             unitary() | q[:4]

Reviewer: Still find the switching of signatures here weird

Author: This is a result of the `operation` decorator. I added this to the tutorial originally a couple of months ago, as several people were asking for the best shortcut to only define the unitary once - worth removing?

Reviewer: No, but worth rethinking the `operation` decorator. Possible to make it something like a `property` which can be called without brackets?

Author: Hmm, I don't think so. It can actually accept (optional) arguments if required, so you can do something like this:

```python
@operation
custom_op(a, b, q):
    Dgate(a) | q[0]
    Rgate(b) | q[1]

with prog.context as q:
    custom_op(a, b) | q[0, 1]
```

However, the `q` argument always disappears from the signature, as the decorator moves it to the right of the `|` operator.

Reviewer: Possible to put the `q` as an argument to the decorator rather than the function?

Author: I'll make an issue and we can look into it :+1:

Author: Ah, started writing out the issue, and released it's not possible that way. The function is a _definition_, `q` is an undefined dummy variable, whereas in the decorator you are calling an already defined function with a `q` that must be defined.

            unitary() | q[4:]
 
            ops.MeasureFock() | q

    Refer to the :class:`~.operation` documentation for more details.

Managing submitted jobs
-----------------------

In the above example, :meth:`RemoteEngine.run` is a **blocking** method; Python will wait until
the job result has been returned from the cloud before further lines
will execute.

> **Non-blocking jobs** can also be submitted using the :meth:`eng.run_async() <~.RemoteEngine.run_async>` method. Unlike ``eng.run()``, this method returns a :class:`~.Job`

Reviewer: Suggest rewording. It's not really the job that's blocking here, it's the call

Author: Rephrased to:
```rest
Jobs can also be submitted using the **non-blocking**
:meth:`eng.run_async() <.RemoteEngine.run_async>` method.
```

>>> job.refresh()
>>> job.status
"complete"
>>> job.result
[[0 1 0 2 1 0 0 0]]

Finally, an incomplete job can be *cancelled* by calling :meth:`job.cancel() <~.Job.cancel>`.

Hardware compilation
--------------------

When creating a quantum program to run on hardware, Strawberry Fields can compile
any collection of the following gates into a multi-mode unitary:

* `General beamsplitters <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.BSgate.html>`_ (:class:`~.ops.BSgate`),

* `Mach-Zehnder interferometers <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.MZgate.html>`_ (:class:`~.ops.MZgate`), or

* `rotations/phase shifts <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.Rgate.html>`_ (:class:`~.ops.Rgate`).

Furthermore, several automatic decompositions are supported:

* You can use the :class:`~.ops.Interferometer` command to directly pass a
  unitary matrix to be decomposed and compiled to match the device architecture.
  This performs a rectangular decomposition using Mach-Zehnder interferometers.

* You can use :class:`~.ops.BipartiteGraphEmbed` to embed a bipartite graph on

>   the GBS chip. Note, however, that the decomposed squeezing values depends on the graph

Reviewer: First time GBS is mentioned, I believe

Author: Ah, good catch. I've replaced it with 'photonic chip' for now. This tutorial is a bit lacking wrt the wider context at the moment, as we do not yet have (a) a introduction page to our photonic hardware, and (b) an introduction to GBS (existing page is aimed more at the advanced level)

>   the GBS chip. Note, however, that the decomposed squeezing values depends on the graph

Reviewer: ```suggestion
  the GBS chip. Note, however, that the decomposed squeezing values depend on the graph
```

>   the GBS chip. Note, however, that the decomposed squeezing values depends on the graph

Reviewer: better to actually make a callout Note here

Author: Made a warning box 🙂 
![image](https://user-images.githubusercontent.com/2959003/77224914-24942880-6bba-11ea-973c-d99f71d98375.png)

  structure, so only bipartite graphs that result in equal squeezing on all
  modes can be executed on currently available chips.

Before sending the program to the cloud platform to be executed, however, Strawberry Fields
must **compile** the program to match the physical architecture or layout of the photonic chip.

> This happens implicitly when using the remote engine, however we can use the :meth:`~.Program.compile`

Reviewer: Does the engine recompile already-compiled programs?

Author: This is the `eng.run` logic:
```python
# if the program hasn't been compiled for this backend, do it now
target = self.backend.circuit_spec
if target is not None and p.target != target:
    p = p.compile(target, **compile_options)
```

So compilation only occurs if the existing program target does not match the engine backend.

        def unitary(q):
            ops.Interferometer(U) | q
            ops.BSgate(0.543, 0.123) | (q[2], q[0])
            ops.Rgate(0.453) | q[1]
            ops.MZgate(0.65, -0.54) | (q[2], q[3])

        prog = sf.Program(8)

        with prog.context as q:
            ops.S2gate(1.0) | (q[0], q[4])
            ops.S2gate(1.0) | (q[1], q[5])
            ops.S2gate(1.0) | (q[3], q[7])

            unitary() | q[:4]
            unitary() | q[4:]

            ops.MeasureFock() | q

    Refer to the :class:`~.operation` documentation for more details.

Managing submitted jobs
-----------------------

In the above example, :meth:`eng.run() <.RemoteEngine.run>` is a **blocking** method; Python will wait until
the job result has been returned from the cloud before further lines
will execute.

Jobs can also be submitted using the **non-blocking** :meth:`eng.run_async() <.RemoteEngine.run_async>`

> method. Unlike ``eng.run()``, which returns a :class:`~.Results` object once the computation is

Reviewer 2: ```suggestion
method. Unlike :meth:`eng.run() <.RemoteEngine.run>`, which returns a :class:`~.Results` object once the computation is
```

            unitary() | q[:4]
            unitary() | q[4:]

            ops.MeasureFock() | q

    Refer to the :class:`~.operation` documentation for more details.

Managing submitted jobs
-----------------------

> In the above example, :meth:`eng.run <.RemoteEngine.run>` is a **blocking** method; Python will wait until

Reviewer: ```suggestion
In the above example, :meth:`eng.run() <.RemoteEngine.run>` is a **blocking** method; Python will wait until
```

        0.06065372-0.11292765j, 0.54733962+0.1215551j, -0.50721513+0.56195975j, -0.15923161+0.26606674j
        0.42212573-0.53182417j, -0.2642572+0.50625182j, 0.19448705+0.28321781j, 0.30281396-0.05582391j
        0.43097587-0.30288974j, 0.07419772-0.21155126j, 0.28335618-0.13633175j, -0.75113453+0.09580304j

    complex array A1[4, 4] =
        -0.13879438-0.47517904j, -0.29303954-0.47264099j, -0.43951987+0.12977568j, -0.03496718-0.48418713j
        0.06065372-0.11292765j, 0.54733962+0.1215551j, -0.50721513+0.56195975j, -0.15923161+0.26606674j
        0.42212573-0.53182417j, -0.2642572+0.50625182j, 0.19448705+0.28321781j, 0.30281396-0.05582391j
        0.43097587-0.30288974j, 0.07419772-0.21155126j, 0.28335618-0.13633175j, -0.75113453+0.09580304j

    # Initial states are two-mode squeezed states
    S2gate(1.0, 0.0) | [0, 4]
    S2gate(1.0, 0.0) | [1, 5]
    S2gate(1.0, 0.0) | [3, 7]

    # Apply the unitary matrix above to
    # the first pair of modes, as well
    # as a beamsplitter
    Interferometer(A0) | [0, 1, 2, 3]
    BSgate(0.543, 0.123) | [2, 0]
    Rgate(0.453) | 1
    MZgate(0.65, -0.54) | [2, 3]

    # Duplicate the above unitary for
    # the second pair of modes

>     Interferometer(A1) | [4, 5, 6, 7]

Reviewer 2: Interesting that it prefers to use another identical matrix `A1` instead of going for `A0` again.

Author: Yep, I didn't know this myself until now!

    BSgate(0.543, 0.123) | [6, 4]
    Rgate(0.453) | 5
    MZgate(0.65, -0.54) | [6, 7]

    # Perform a PNR measurement in the Fock basis
    MeasureFock() | [0, 1, 2, 3, 4, 5, 6, 7]

After you have created your Blackbird script, either by exporting it from
Strawberry Fields or writing it by hand, it can be remotely executed using the command line,

.. code-block:: console

    $ sf run program1.xbb --output out.txt

After executing the above command, the result will be stored in ``out.txt`` in the
current working directory. You can also omit the ``--output`` parameter to print the
result to the screen.

Furthermore, saved Blackbird scripts can be imported as a Strawberry Fields program
using the :func:`~.load` function:

>>> prog = load("test.xbb")

Embedding bipartite graphs
--------------------------

The currently available hardware devices support embedding bipartite graphs,
i.e., those with adjacency matrices

.. math:: A = \begin{bmatrix}0 & B\\ B^T & 0\end{bmatrix}

where :math:`B` represents the edges between the two sets of
vertices in the graph. However, the devices are currently restricted

> to bipartite graphs such that the singular values form the set :math:`\{0, d\}`

Reviewer 2: ```suggestion
to bipartite graphs with equally sized partitions, such that the singular values form the set :math:`\{0, d\}`
```
Not sure though :thinking:

for some real value :math:`d`.

Here, we will
consider a `complete bipartite graph <https://en.wikipedia.org/wiki/Complete_bipartite_graph>`_,

> since the singular values are of the form :math:`\{d, 0\}`.

Reviewer 2: ```suggestion
since the singular values are of the form :math:`\{0, d\}`.
```
Just to match the one above.

Rgate(-1.173) | (q[5])
MeasureFock | (q[0], q[1], q[2], q[3], q[4], q[5], q[6], q[7])

While equivalent to the uncompiled program, we can now see the low-level hardware
operations that are applied on the physical chip.

Working with Blackbird scripts
------------------------------

When submitting quantum programs to be executed remotely, they are communicated to

> the cloud platform using Blackbird --- a quantum photonic assembly language.

Reviewer: ```suggestion
the cloud platform using Blackbird---a quantum photonic assembly language.
```

Strawberry Fields also supports exporting programs directly as Blackbird scripts
(an ``xbb`` file); Blackbird scripts can then be submitted to be executed via the

> Strawberry Fields :doc:`command line interface </code/sf_cli>`.

Reviewer: Does this link render properly? (didn't check myself, but looks similar to ones @trbromley commented on)

Author: Yep! 
![image](https://user-images.githubusercontent.com/2959003/77225480-67a4ca80-6bbf-11ea-8827-a1df10ae7743.png)

For example, lets save our program above as a Blackbird script in the
current directory, via the :func:`~.save` function:

>>> sf.save("program1.xbb", prog)

This produces the following Blackbird script, where comments have been added

> for clarity:

Reviewer: ```suggestion
This produces the following Blackbird script (comments have been added
for clarity):
```

        0.43097587-0.30288974j, 0.07419772-0.21155126j, 0.28335618-0.13633175j, -0.75113453+0.09580304j

    # Initial states are two-mode squeezed states
    S2gate(1.0, 0.0) | [0, 4]
    S2gate(1.0, 0.0) | [1, 5]
    S2gate(1.0, 0.0) | [3, 7]

    # Apply the unitary matrix above to
    # the first pair of modes, as well
    # as a beamsplitter
    Interferometer(A0) | [0, 1, 2, 3]
    BSgate(0.543, 0.123) | [2, 0]
    Rgate(0.453) | 1
    MZgate(0.65, -0.54) | [2, 3]

    # Duplicate the above unitary for
    # the second pair of modes
    Interferometer(A1) | [4, 5, 6, 7]
    BSgate(0.543, 0.123) | [6, 4]
    Rgate(0.453) | 5
    MZgate(0.65, -0.54) | [6, 7]

    # Perform a PNR measurement in the Fock basis
    MeasureFock() | [0, 1, 2, 3, 4, 5, 6, 7]

After you have created your Blackbird script, either by exporting it from
Strawberry Fields or writing it by hand, it can be remotely executed using the command line,

.. code-block:: console

    $ sf run program1.xbb --output out.txt

After executing the above command, the result will be stored in ``out.txt`` in the
current working directory. You can also omit the ``--output`` parameter to print the
result to the screen.

> Furthermore, saved Blackbird scripts can be imported as a Strawberry Fields program

Reviewer: ```suggestion
Furthermore, saved Blackbird scripts can be imported as Strawberry Fields programs
```

using the :func:`~.load` function:

>>> prog = load("test.xbb")

Embedding bipartite graphs
--------------------------

The currently available hardware devices support embedding bipartite graphs,
i.e., those with adjacency matrices

.. math:: A = \begin{bmatrix}0 & B\\ B^T & 0\end{bmatrix}

where :math:`B` represents the edges between the two sets of
vertices in the graph. However, the devices are currently restricted
to bipartite graphs with equally sized partitions, such that the singular values form the set :math:`\{0, d\}`
for some real value :math:`d`.

Here, we will
consider a `complete bipartite graph <https://en.wikipedia.org/wiki/Complete_bipartite_graph>`_,

> since the singular values are of the form :math:`\{0, d\}`.

Reviewer: Didn't like my `zeros_like` suggestion? :crying_cat_face:

Author: Oh, I wasn't sure what you meant by it! And then in the process of making the new PR, accidentally lost track of that comment. Did you mean write `A` using `zeros_like`?

Reviewer: Instead of using `0*B`, use `np.zeros_like(B)`

Rgate(-2.51) | (q[5])
MeasureFock | (q[0], q[1], q[2], q[3], q[4], q[5], q[6], q[7])

If the bipartite graph to be embedded does not satisfy the aforementioned
restriction on the singular values, an error message will be raised on
compilation:

>>> B = np.array([[0, 1, 0, 1], [1, 0, 1, 0], [0, 1, 1, 1], [1, 0, 1, 0]])
>>> A = np.block([[0*B, B], [B.T, 0*B]])
>>> prog = sf.Program(8)
>>> with prog.context as q:
...     ops.BipartiteGraphEmbed(A, mean_photon_per_mode=1) | q
...     ops.MeasureFock() | q
CircuitError: Incorrect squeezing value(s) (r, phi)={(1.336, 0.0), (0.177, 0.0), (0.818, 0.0)}.

> Allowed squeezing value(s) are (r, phi)={(1, 0.0), (0.0, 0.0)}.

Reviewer 2: I'm not sure if this is that useful for the tutorial. If a user did run into the error message, it is quite self explanatory.

Rgate(-2.51) | (q[5])
MeasureFock | (q[0], q[1], q[2], q[3], q[4], q[5], q[6], q[7])

If the bipartite graph to be embedded does not satisfy the aforementioned
restriction on the singular values, an error message will be raised on
compilation:

>>> B = np.array([[0, 1, 0, 1], [1, 0, 1, 0], [0, 1, 1, 1], [1, 0, 1, 0]])
>>> A = np.block([[0*B, B], [B.T, 0*B]])
>>> prog = sf.Program(8)
>>> with prog.context as q:
...     ops.BipartiteGraphEmbed(A, mean_photon_per_mode=1) | q
...     ops.MeasureFock() | q
CircuitError: Incorrect squeezing value(s) (r, phi)={(1.336, 0.0), (0.177, 0.0), (0.818, 0.0)}.

> Allowed squeezing value(s) are (r, phi)={(1, 0.0), (0.0, 0.0)}.

Reviewer: does the error message actually print (1, 0.0)? i.e., a combination of integer and float?

Author: Just double checked, it does. This is an artifact of the allowed squeezing value being an integer `1` in the chip spec, which is being programmatically inserted into the error message. We should fix this, but better to do it in a separate PR once the new HW chips have been added

Reviewer: k can we document so we don't forget?

Rgate(-2.51) | (q[5])
MeasureFock | (q[0], q[1], q[2], q[3], q[4], q[5], q[6], q[7])

If the bipartite graph to be embedded does not satisfy the aforementioned
restriction on the singular values, an error message will be raised on
compilation:

>>> B = np.array([[0, 1, 0, 1], [1, 0, 1, 0], [0, 1, 1, 1], [1, 0, 1, 0]])

> >>> A = np.block([[0*B, B], [B.T, 0*B]])

Author: ```suggestion
>>> A = np.block([[np.zeros_like(B), B], [B.T, np.zeros_like(B)]])
```
